### ISSUE TITLE
Appending to file with astropy.io.ascii.write()

### ISSUE THREAD

**Issue by rrgupta** [astropy/astropy#3684]
I am using astropy v0.4.2 and would like to know if there is a way to append to a file when using `astropy.io.ascii.write()` similar to the append in Python:

```
f = open('filename.txt', 'a')
f.write("appended text")
```

Perhaps this functionality already exists, but I cannot find documentation or examples of it. It looks like it exists for HDF5 files (http://astropy.readthedocs.org/en/latest/io/unified.html#hdf5) but not ASCII. 
If appending is currently not supported for `astropy.io.ascii.write()`, I would like to request that it be added, since it would be very useful.
Thanks!

**Comment by embray** [comment]
I don't think `file.write` in the standard lib is really analogous.  But in any case you can write to a file handled opened in append mode to append to an existing table:

```python
In [1]: from astropy.table import Table

In [2]: import os

In [3]: t1 = Table({'a': [1, 2, 3], 'b': [4, 5, 6]}, names=('a', 'b'))

In [4]: t1.write('table.txt', format='ascii')

In [5]: t2 = Table({'a': [7, 8, 9], 'b': [10, 11, 12]}, names=('a', 'b'))

In [6]: cat table.txt
a b
1 4
2 5
3 6

In [7]: with open('table.txt', mode='a') as f:
   ...:     f.seek(0, os.SEEK_END)  # Some platforms don't automatically seek to end when files opened in append mode
   ...:     t2.write(f, format='ascii.no_header')
   ...:     

In 8]: cat table.txt
a b
1 4
2 5
3 6
7 10
8 11
9 12
```

**Comment by rrgupta** [comment]
Thanks, that does what I wanted! You're right, it's not really analogous to Python's standard `file.write`, but I guess what I was wondering was if there was something more analogous to the HDF5 case I referenced where you can just add the keyword argument `append=True` that would work for `astropy.io.ascii.write()`.  I'm guessing the answer is no, then?   This workaround gets the job done but having this keyword available would certainly be simpler.

**Comment by embray** [comment]
Yeah, like I said if I had to guess the main complication is just that the `Table` class itself doesn't carry much information, I don't think, about how its source file was formatted (though someone with more knowledge about that can correct me).  So it's easy to screw up appending new rows to an ASCII file if those rows aren't already formatted consistently with the existing file.  That is, just saying `Table.write(..., append=True)` has no obvious way of working for ASCII tables unless the format is already known or can be guessed.  Whereas for something like HDF5 you're just updating an existing data structure with a known format that can easy be checked to match the table you're trying to append to it.  This would also work better with "standardized" ASCII formats like [ECSV](https://github.com/astropy/astropy-APEs/blob/master/APE6.rst)

**Comment by rrgupta** [comment]
Right, good point....OK then, I'm satisfied.  Thanks for the help!

**Comment by embray** [comment]
Reopening in case anyone else wants to comment; I think there are some possibilities here, just not sure exactly what's practical.

**Comment by taldcroft** [comment]
@embray - a simple, reliable, but inefficient implementation would be to have the `write` process internally read the existing table, then `vstack` the existing and new tables (which can do all the validation to make sure things line up exactly with `join_type='exact'`), then rewrite the stacked table.  As I said, it's quite inefficient but would leverage existing functionality to make things reliable and provide what amounts to a convenience method for users.

Trying to be more clever and not re-write the original table introduces potential complications for some formats, particularly any that have fixed width output.

As you hinted, ECSV is somewhat unique in providing all the table meta upfront in a way that could potentially allow an append without re-write.  Even that is not entirely without complication since `vstack` can upgrade a column type, e.g. if you stack int8 and int64 it turns to int64, so the original `column_type` would need fixing.  Rewriting the header for ECSV would get sticky in that case.

**Comment by embray** [comment]
I don't think the original table needs to even be read into memory though.  Perhaps it could be *parsed*, but the results immediately thrown away except the bare minimum to determine what the data types of the columns should be.  Then whatever you're appending would have to match up exactly with what was already there.  This would definitely have some limitations.  For example if there is a column that is only 3 characters wide and all the values in it were < 255 then the maximum viable data type for that column would be uint8, say.  Then any data being append to that column would have to be restricted to that.

So to summarize, I think it's doable without reading the original table into memory (it would still have to be parsed line by line though), so long as there are strong restrictions on what can be appended.

**Comment by taldcroft** [comment]
@embray - sure, that's all possible but it becomes a major project with lots of new code and known limitations that mean more complicated testing, docs, and so on.  I was  suggesting something that we could do today in a half a page of code (maybe...) with very few limitations.

**Comment by hamogu** [comment]
I think this is too complicated in practice. As @taldcroft said we would be adding a lot of code and a lot of tests for a case that probably doesn't come up too often. instead I suggest adding the wirkaround to the documentation as an example so that users can use that if they want to. If its done explicitly like that, then the burden to ensure that the format matches up is on the user, and he or she presumably knows exactly what the file looks like.

**Comment by embray** [comment]
I don't think it's *too* complicated; just not worth the effort for anyone involved right now--definitely low priority.  Why not just leave this open for someone who would like to tackle it, if ever?

**Comment by embray** [comment]
(In the meantime an example in the docs of how to do this with vstack + join + the workaround I suggested would make sense)

**Comment by taldcroft** [comment]
Closing, no action in 6 years. The original idea of a one-line way of appending an ASCII file has been rejected. There is a possible doc fix suggested in the conversation but I think we can let that go.
